mne/viz/utils.py

```diff
@@ -386,3 +385,119 @@ def figure_nobar(*args, **kwargs):
+    Display an image so you can click on it and store x/y positions.
+
+    Takes as input an image array (can be any array that works with imshow,
+    but will work best with images.  Displays the image and lets you
+    click on it.  Stores the xy coordinates of each click, so now you can
+    superimpose something on top of it.
+
+    Upon clicking, the x/y coordinate of the cursor will be stored in
+    self.coords, which is a list of (x, y) tuples.
+
+    Inputs
+    ------
+    imdata: ndarray
+        The image that you wish to click on for 2-d points.
+
+    Notes
+    -----
+        .. versionadded:: 0.9.0
+
+    """
+
+    def __init__(self, imdata, **kwargs):
+        """Display the image for clicking."""
+        from matplotlib.pyplot import figure, show
+        self.coords = []
+        self.imdata = imdata
+        self.fig = figure()
+        self.ax = self.fig.add_subplot(111)
+        self.ymax = self.imdata.shape[0]
+        self.xmax = self.imdata.shape[1]
+        self.im = self.ax.imshow(imdata, aspect='auto',
+                                 extent=(0, self.xmax, 0, self.ymax),
+                                 picker=True, **kwargs)
+        self.ax.axis('off')
+        self.fig.canvas.mpl_connect('pick_event', self.onclick)
+        show()
+
+    def onclick(self, event):
+        """Mouse click handler."""
+        mouseevent = event.mouseevent
+        self.coords.append((mouseevent.xdata, mouseevent.ydata))
+
+    def plot_clicks(self, **kwargs):
+        """Plot the x/y positions stored in self.coords."""
+        from matplotlib.pyplot import subplots, show
+        f, ax = subplots()
+        ax.imshow(self.imdata, extent=(0, self.xmax, 0, self.ymax), **kwargs)
+        xlim, ylim = [ax.get_xlim(), ax.get_ylim()]
+        xcoords, ycoords = zip(*self.coords)
+        ax.scatter(xcoords, ycoords, c='r')
+        ann_text = np.arange(len(self.coords)).astype(str)
+        for txt, coord in zip(ann_text, self.coords):
+            ax.annotate(txt, coord, fontsize=20, color='r')
+        ax.set_xlim(xlim)
+        ax.set_ylim(ylim)
+        show()
+
+    def to_layout(self, **kwargs):
+        """Turn coordinates into an MNE Layout object.
+
+        Normalizes by the image you used to generate clicks
+        """
+        from mne.channels.layout import generate_2d_layout
+        coords = np.array(self.coords)
+        lt = generate_2d_layout(coords, bg_image=self.imdata, **kwargs)
+        return lt
+
+
+def _fake_click(fig, ax, point, xform='ax'):
+    """Helper to fake a click at a relative point within axes."""
+    if xform == 'ax':
+        x, y = ax.transAxes.transform_point(point)
+    elif xform == 'data':
+        x, y = ax.transData.transform_point(point)
+    else:
+        raise ValueError('unknown transform')
+    try:
+        fig.canvas.button_press_event(x, y, 1, False, None)
+    except:  # for old MPL
+        fig.canvas.button_press_event(x, y, 1, False)
+
+
+def add_background_image(fig, im):
```

Comment:
Mmk, how's that? Also added a test.

On Mon, Apr 27, 2015 at 12:21 PM, Eric Larson <notifications@github.com>
wrote:

> In mne/viz/utils.py
> <https://github.com/mne-tools/mne-python/pull/1762#discussion_r29180898>:
>
> > +
> > +def _fake_click(fig, ax, point, xform='ax'):
> > +    """Helper to fake a click at a relative point within axes."""
> > +    if xform == 'ax':
> > +        x, y = ax.transAxes.transform_point(point)
> > +    elif xform == 'data':
> > +        x, y = ax.transData.transform_point(point)
> > +    else:
> > +        raise ValueError('unknown transform')
> > +    try:
> > +        fig.canvas.button_press_event(x, y, 1, False, None)
> > +    except:  # for old MPL
> > +        fig.canvas.button_press_event(x, y, 1, False)
> > +
> > +
> > +def add_background_image(fig, im):
>
> adding a set_ratios flag is fine by me
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762/files#r29180898>.
>